pkg/relay/ethereum/error_parse.go

```diff
@@ -0,0 +1,113 @@
+package ethereum
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+)
+
+type ErrorRepository map[[4]byte]abi.Error
+
+func CreateErrorRepository(abiPaths []string) (ErrorRepository, error) {
+	var errABIs []abi.Error
+
+	for _, dir := range abiPaths {
+		if err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() && strings.HasSuffix(info.Name(), ".json") {
+				data, err := os.ReadFile(path)
+				if err != nil {
+					return err
+				}
+
+				contractABI, err := abi.JSON(bytes.NewReader(data))
+				for _, error := range contractABI.Errors {
+					errABIs = append(errABIs, error)
+				}
+			}
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	return NewErrorRepository(errABIs)
+}
+
+func NewErrorRepository(errABIs []abi.Error) (ErrorRepository, error) {
+	repo := make(ErrorRepository)
+	for _, errABI := range errABIs {
+		if err := repo.Add(errABI); err != nil {
+			return nil, err
+		}
+	}
+
+	return repo, nil
+}
+
+func (r ErrorRepository) Add(errABI abi.Error) error {
+	var sel [4]byte
+	copy(sel[:], errABI.ID[:4])
+	if existingErrABI, ok := r[sel]; ok {
+		if existingErrABI.Sig == errABI.Sig {
+			return nil
+		}
+		return fmt.Errorf("error selector collision: selector=%x, newErrABI=%v, existingErrABI=%v", sel, errABI, existingErrABI)
+	}
+	r[sel] = errABI
+	return nil
+}
+
+func (r ErrorRepository) Get(errorData []byte) (abi.Error, error) {
+	if len(errorData) < 4 {
+		return abi.Error{}, fmt.Errorf("the size of error data is less than 4 bytes: errorData=%x", errorData)
+	}
+	var sel [4]byte
+	copy(sel[:], errorData[:4])
+	errABI, ok := r[sel]
+	if !ok {
+		return abi.Error{}, fmt.Errorf("error ABI not found")
+	}
+	return errABI, nil
+}
+
+func errorToJSON(errVal interface{}, errABI abi.Error) (string, error) {
```

Comment:
I think type of `errVal` should be `[]interface{}` more type-safe

Reply:
I think it doesn't matter which of a caller and a callee executes type assertion.
Do you mean that we should use type assertion that returns two values (panic-free) instead of one value (panicable)?
If so, I agree with it. Especially because this code is only for logging, it should be panic-free.

Reply:
My nit comment means that the ` errVal` parameter of ` errorToJSON` function always panic if the argument is not of type `[]interface{}`, so it is better to set the paramter type to `[]interface{}`. However maybe we could just use type assertion with two return values in `errorToJSON` instead.

Reply:
Thank you. Fixed https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/39/commits/7d9990a9742112325da8a36877538d6f06dac754